### PR TITLE
Share url implementation

### PR DESCRIPTION
**Pull request by goyal1092**
## Description

URL updates when filters are changed.

## Related Issue
https://wazimap.atlassian.net/browse/WNCM-638

## How should a reviewer test it locally
Changing filters should update url
Copy pasting that url should apply same filters
applied filters should come in my view
Pressing forward and backward on browser should apply/remove filters
url with geography should open up specific geography with filters applied

## Checklist

- [ ]  🚀 is the code ready to be merged and go live?
- [ ]  🛠 does it work (build) locally
- [ ] 👩‍🎨 does the design match the [Demo](https://wazimap-ng-v1.webflow.io/demo)

### Pull Request

- [ ]  📰 good title
- [ ]  📝good description
- [ ]  🔖 issue linked
- [ ]  📖 changelog filled out
- [ ] commit messages are meaningful

### Code Quality

- [ ]  🚧 no commented out code
- [ ]  🖨 no unnecessary logging
- [ ]  🎱 no magic numbers

### Testing

- [ ]  ✅ added (appropriate) unit tests
- [ ]  💢 edge cases in tests were considered
- [ ]  ✅ ran tests locally & are passing

### CODE REVIEW

File: src/js/models/data_filter_model.js [hunk 1]
```diff
@@ -28,14 +28,14 @@ export class DataFilterModel extends Observable {
         points: 'points'
     }
 
-    constructor(groups, configFilters, previouslySelectedFilters, primaryGroup, childData, filterType = DataFilterModel.FILTER_TYPE.indicators) {
```
Comment (emre2038) [inline src/js/models/data_filter_model.js:31]: can't we have 1 single property for `previouslySelectedFilters` and `urlFilters`? do we need them separated?

File: src/js/profile/chart.js [hunk 2]
```diff
@@ -440,7 +456,7 @@ export class Chart extends Component {
     };
 
     handleChartFilter = (indicators, groups) => {
```
Comment (emre2038) [inline src/js/profile/chart.js:458]: same thing here, do we need `previouslySelectedFilters` and `urlParamFilters` separated?

File: src/js/load.js [hunk 3]
```diff
@@ -110,7 +110,8 @@ class Application extends Component {
 
         controller.on('profile.loaded', payload => {
```
Comment (emre2038) [inline src/js/load.js:111]: can we move this into one of the js files in the `setup` folder

File: src/js/ui_components/filter_row.js [hunk 4]
```diff
@@ -259,7 +275,7 @@ export class FilterRow extends Component {
         this.model.currentIndicatorValue = selectedItem;
     }
 
-    onSubindicatorSelected(selectedItem) {
+    onSubindicatorSelected(selectedItem, manualTrigger) {
```
Comment (emre2038) [inline src/js/ui_components/filter_row.js:278]: can we remove `manualTrigger`?

File: src/js/ui_components/filter_row.js [hunk 5]
```diff
@@ -284,7 +300,6 @@ export class FilterRow extends Component {
     removeRow() {
         this.model.currentIndicatorValue = null;
         $(this.container).remove();
-        this.model.dataFilterModel.updateFilteredData();
```
Comment (emre2038) [inline src/js/ui_components/filter_row.js:287]: didn't we need this?

File: src/js/load.js (hunk 3 above)
Comment (goyal1092) [reply src/js/load.js:111]: Should i make a new file for this @emre2038 
If yes what should i name this ?

File: src/js/ui_components/filter_row.js (hunk 4 above)
Comment (goyal1092) [reply src/js/ui_components/filter_row.js]: Done

File: src/js/ui_components/filter_row.js (hunk 5 above)
Comment (goyal1092) [reply src/js/ui_components/filter_row.js:287]: yup .. Done

File: __tests__/gui/sharing/sharing.js [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    })
+
+    cy.visit('/#geo:WC');
+})
+
+Then('I wait until map is ready for Western Cape', () => {
+    waitUntilGeographyIsLoaded('Western Cape');
+})
+
+When('I go back to root geograohy', () => {
```
Comment (michaelglenister) [inline __tests__/gui/sharing/sharing.js:241]: Typo

File: __tests__/gui/sharing.feature [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    And I collapse Rich Data Panel
+
+    When I expand My View Window
+    And I click on "INDICATOR OPTIONS" in My View
+    Then I confirm that there is an indicator filter for "Data mapper:Region of birth:age:30-35" at index 0
+    Then I confirm that there is an indicator filter for "Data mapper:Region of birth:language:English" at index 1
+    Then I confirm that there is an indicator filter for "Rich data view:Language most spoken at home:language:Afrikaans" at index 2
+    Then I collapse My View Window
+
+    When I go back to root geograohy
```
Comment (michaelglenister) [inline __tests__/gui/sharing.feature:213]: Typo

File: src/js/controller.js [hunk 8]
```diff
@@ -234,6 +254,97 @@ export default class Controller extends Component {
...
+          }
+        }
+      } else if (selectedFilters.length === 0 && profileView !== null) {
+        history.pushState(
+          {
+            "filters": this._filteredIndicators,
+            "profileView": currentState
+          }, '', `/${window.location.hash}`
+        );
+      }
```
Comment (michaelglenister) [inline src/js/controller.js:306]: It looks like the first two pushState are identical and the third pushState is very similar. That makes be think some of these if statements are not needed or there can be variable to help reduce the repeated code like

```
let stateParams = {...}

if (...) {
  pushState(stateParams, "location1");
}
else {
  pushState(stateParams, "location2");
}
```

File: src/js/load.js (hunk 3 above)
Comment (michaelglenister) [reply src/js/load.js:111]: Maybe a good place for this is https://github.com/OpenUpSA/wazimap-ng-ui/blob/staging/src/js/setup/general.js

**Review by michaelglenister — COMMENTED** [review]
Looks good so far, mainly one suggestion to clean up if statements controlling how pushState is used in [src/js/controller.js](https://github.com/OpenUpSA/wazimap-ng-ui/pull/682/files#diff-f47c60f82e0b188cc72db120e2829c4473753d20186183b4c2e1492169d61321)

**Review by michaelglenister — APPROVED** [review]
Thanks, the changes to pushState() in controller.js made it easier to understand

**Pull request merged**